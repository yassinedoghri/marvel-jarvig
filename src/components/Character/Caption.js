import { transitions } from "polished/lib/index";
import styled from "styled-components";

const Caption = styled.figcaption`
  position: absolute;
  width: ${props => (props.checked ? "100%" : "auto")};
  bottom: ${props => (props.checked ? "-.5em" : "1em")};
  left: ${props => (props.checked ? "0" : "1em")};
  font-weight: ${props => (props.checked ? "normal" : "900")};
  text-align: center;
  font-size: ${props => props.theme.fontSize.sm}rem;
  background: ${props => props.theme.colors.foreground.base};
  color: ${props => props.theme.colors.contrastDark.primary};
  padding: 0.2em 0.5em;
  border: solid ${props => props.theme.colors.foreground.base};
  border-width: 2px 2px 3px 3px;
  border-radius: ${props =>
    props.checked ? "4% 95% 6% 95%/95% 4% 92% 5%" : "25px"};
  transform: rotate(-2deg);
  ${transitions("all 200ms ease-in-out")};
`;

export default Caption;
