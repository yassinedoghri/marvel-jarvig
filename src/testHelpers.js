import React from "react";
import "jest-styled-components";
import renderer from "react-test-renderer";
import theme from "theme";
import { ThemeProvider } from "styled-components";

export default function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}
