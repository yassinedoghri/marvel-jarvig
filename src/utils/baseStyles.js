import { normalize } from "polished";
import { injectGlobal } from "styled-components";

export default () => injectGlobal`
    ${normalize()}

    @import url("https://fonts.googleapis.com/css?family=Bangers");
    @import url("https://fonts.googleapis.com/css?family=Roboto:400,900");
    @import url("https://fonts.googleapis.com/css?family=Press+Start+2P");
  
    * {
      font-family: "Roboto", sans-serif;
      box-sizing: border-box;
    }
    
    *:focus {
      outline: none;
    }
    
    a {
      text-decoration: none;
    }
    
    ul li, ol li {
        list-style: none;
    }
    
    html, body, #root {
      height: 100%;
    }
    
    body {
      padding: 0;
      background-color: white;
    }
`;
