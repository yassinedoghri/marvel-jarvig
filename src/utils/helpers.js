export const randomIntFromInterval = (min, max) =>
  Math.floor(Math.random() * (max - min + 1) + min);

export const getArrayCounts = arr => {
  const counts = {};

  for (let i = 0; i < arr.length; i += 1) {
    const num = arr[i];
    counts[num] = counts[num] ? counts[num] + 1 : 1;
  }
  return counts;
};

export const generateQuery = params => {
  const esc = encodeURIComponent;
  return Object.keys(params)
    .map(k => `${esc(k)}=${esc(params[k])}`)
    .join("&");
};

export const generateAnswers = (
  characters,
  charactersPerQuestion,
  bannedChars
) => {
  const charactersToDisplay = [];
  const visitedIndices = [];
  while (charactersToDisplay.length < charactersPerQuestion) {
    const randomItem = randomIntFromInterval(0, 99);
    const currChar = characters[randomItem];
    if (!visitedIndices.includes(randomItem)) {
      // remove duplicates: check if character has been already added
      visitedIndices.push(randomItem); // add it to visited indices if not
      if (
        !currChar.thumbnail.path.includes("image_not_available") &&
        !bannedChars.includes(currChar.name)
      ) {
        // check if character has image and not in banned characters
        charactersToDisplay.push(currChar); // only add characters with image
      }
    }
  }
  const answer =
    charactersToDisplay[randomIntFromInterval(0, charactersPerQuestion - 1)];

  return {
    charactersToDisplay,
    answer
  };
};
