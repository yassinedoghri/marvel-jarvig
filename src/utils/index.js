export { default as style } from "utils/style";
export { default as baseStyles } from "utils/baseStyles";
export { default as helpers } from "utils/helpers";
export { default as testHelpers } from "utils/testHelpers";
export { default as registerServiceWorker } from "utils/registerServiceWorker";
export { default as media } from "utils/media";
