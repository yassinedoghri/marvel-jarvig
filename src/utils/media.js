import { css } from "styled-components";

const sizes = {
  giant: 1600,
  desktop: 1024,
  tablet: 768,
  phone: 480
};

// iterate through the sizes and create a media template
const media = Object.keys(sizes).reduce((accumulator, label) => {
  // use em in breakpoints to work properly cross-browser and support users
  // changing their browsers font-size: https://zellwk.com/blog/media-query-units/
  const emSize = sizes[label] / 16;
  accumulator[label] = (...args) => css`
    @media (min-width: ${emSize}em) {
      ${css(...args)};
    }
  `;
  return accumulator;
}, {});

export default media;
