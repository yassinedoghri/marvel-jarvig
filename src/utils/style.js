export function getColor(theme, color = "", tint = "") {
  let colorSet = color === "" ? "default" : color; // default color
  let tintSet = tint === "" ? "base" : tint; // default tint for base colors
  if (color.startsWith("contrast")) {
    colorSet = color;
    tintSet = tint === "" ? "primary" : tint;
  }
  return theme.colors[colorSet][tintSet];
}

export function setFontSize(theme, size = "md") {
  return `
    font-size: ${theme.fontSize[size]}rem;
  `;
}

export function getResultColor(theme, result) {
  switch (result) {
    case "correct":
      return theme.colors.accent2.base;
    case "incorrect":
      return theme.colors.primary.base;
    default:
      return theme.colors.foreground.base;
  }
}
