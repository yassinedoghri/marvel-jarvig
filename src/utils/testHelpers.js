import { createMemoryHistory } from "history";
import React from "react";
import { Provider } from "react-redux";
import {
  ConnectedRouter,
  routerMiddleware,
  routerReducer
} from "react-router-redux";
import renderer from "react-test-renderer";
import ShallowRenderer from "react-test-renderer/shallow";
import rootReducer from "reducers";
import { applyMiddleware, combineReducers, createStore } from "redux";
import { ThemeProvider } from "styled-components";
import theme from "theme";

const history = createMemoryHistory("/");
const routeMiddleware = routerMiddleware(history);

const store = createStore(
  combineReducers({
    ...rootReducer,
    router: routerReducer
  }),
  applyMiddleware(routeMiddleware)
);

export function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}

export function renderInContext(component) {
  return renderer.create(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={theme}>{component}</ThemeProvider>
      </ConnectedRouter>
    </Provider>
  );
}

export function ShallowRenderInContext(component) {
  const renderer = new ShallowRenderer();
  return renderer.render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={theme}>{component}</ThemeProvider>
      </ConnectedRouter>
    </Provider>
  );
}
