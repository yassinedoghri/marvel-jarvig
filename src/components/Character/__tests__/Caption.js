import Caption from "components/Character/Caption";
import React from "react";
import { renderWithTheme } from "utils/testHelpers";

describe("Character Caption", () => {
  it("renders correctly", () => {
    const tree = renderWithTheme(<Caption>Iron Man</Caption>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("adds the checked modifier", () => {
    const tree = renderWithTheme(<Caption checked>Iron Man</Caption>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
