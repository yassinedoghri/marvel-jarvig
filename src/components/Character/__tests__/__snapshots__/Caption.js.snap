// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Caption adds the checked modifier 1`] = `
<figcaption
  checked={true}
  className="sc-bdVaJa jvulXB"
>
  Iron Man
</figcaption>
`;

exports[`Character Caption renders correctly 1`] = `
<figcaption
  className="sc-bdVaJa eykFkZ"
>
  Iron Man
</figcaption>
`;
